src/main/java/com/commercetools/sync/commons/utils/QueryAll.java

```diff
@@ -1,166 +1,180 @@
+     * then a new future of the next page is returned. If there are no more results, the method returns null.
      *
-     * @param client        the CTP client that the query is run on.
-     * @param totalElements the total number of elements resulting from the query.
-     * @param callback      the callback to apply on each page of results.
-     * @return a list of futures containing the results of applying the callback on each page of results.
+     * @param page the current page result.
+     * @return If there is a next page, then a new future of the next page is returned. If there are no more results,
+     *         the method returns null.
      */
-    @Nonnull
-    private <S> List<CompletableFuture<S>> queryNextPages(final SphereClient client, final long totalElements,
-                                                          final Function<List<T>, S> callback) {
-        final long totalPages = getTotalNumberOfPages(totalElements);
-        return LongStream.range(1, totalPages)
-                         .mapToObj(page -> queryPage(client, page)
-                             .thenApply(PagedQueryResult::getResults)
-                             .thenApply(callback)
-                             .toCompletableFuture())
-                         .collect(toList());
+    @Nullable
+    private CompletionStage<PagedQueryResult<T>> processPageAndGetNext(@Nonnull final PagedQueryResult<T> page) {
+        final List<T> currentPageElements = page.getResults();
+        if (currentPageElements.size() == pageSize) {
+            mapOrConsume(currentPageElements);
+            return getNextPageStage(currentPageElements);
+        } else {
+            if (currentPageElements.size() > 0) {
+                mapOrConsume(currentPageElements);
+            }
+        }
+        return null;
     }
 
     /**
-     * Given a total number of elements {@code totalElements} this method calculates the number of pages needed that
-     * would be needed to cover all the elements with set {@code pageSize} of {@link this} instance.
+     * Given a list of page elements of resource {@code T}, this method checks if this instance's {@code pageConsumer}
+     * or {@code pageMapper} is set (not null). The one which is set is then applied on the list of page elements.
+     *
      *
-     * @param totalElements number of elements to get the number of pages for.
-     * @return the total number of pages.
+     * @param pageElements list of page elements of resource {@code T}.
      */
-    long getTotalNumberOfPages(final long totalElements) {
-        return (long) Math.ceil((double) totalElements / pageSize);
+    private void mapOrConsume(@Nonnull final List<T> pageElements) {
+        if (pageConsumer != null) {
+            pageConsumer.accept(pageElements);
+        } else {
+            mappedResultsTillNow.add(pageMapper.apply(pageElements));
```

Comment:
as we discussed, here the mapper is on the page but not each element in the page. This is decision we need to take on how to apply on the JVM SDK convenience change of QueryAll but not here then. Since this is temporary, as mentioned in the description of this pr, and should be removed from here and use the JVM SDK one instead.